File: frontend/src/lib/cuda.ts
// CUDA version parsing and comparison, mirroring the backend's app/core/cuda.py.
// An offering's min_cuda is compared against an image's cuda_version to hide incompatible images.

function parseCuda(v?: string | null): [number, number] | null {
  if (!v) return null;
  const m = /^(\d+)(?:\.(\d+))?/.exec(String(v).trim());
  if (!m) return null;
  return [Number(m[1]), Number(m[2] ?? 0)];
}

// True when the image's CUDA meets the offering's minimum. Permissive: if either side is unset,
// the pair passes.
export function cudaCompatible(imageCuda?: string | null, minCuda?: string | null): boolean {
  const iv = parseCuda(imageCuda);
  const mv = parseCuda(minCuda);
  if (!iv || !mv) return true;
  return iv[0] > mv[0] || (iv[0] === mv[0] && iv[1] >= mv[1]);
}

File: backend/app/core/cuda.py
"""Parsing and comparison of CUDA versions.

GPU compatibility is determined by the architecture — the compute capability — which translates into
the minimum CUDA toolkit version that supports the card. Offerings carry min_cuda per GPU model and
images carry cuda_version; the pair is compatible only when
image.cuda_version >= offering.min_cuda.
"""
from __future__ import annotations


def parse_cuda(v: str | None) -> tuple[int, int] | None:
    """Parse '12.8' to (12, 8) and '11' to (11, 0). Empty or malformed input yields None."""
    if not v:
        return None
    try:
        parts = str(v).strip().split(".")
        major = int(parts[0])
        minor = int(parts[1]) if len(parts) > 1 and parts[1] != "" else 0
        return (major, minor)
    except (ValueError, IndexError):
        return None


def cuda_compatible(image_cuda: str | None, min_cuda: str | None) -> bool:
    """True when the image's CUDA version meets or exceeds the offering's minimum.

    Deliberately permissive: if either side is unset or malformed the pair passes, so an unknown
    constraint never hides an image.
    """
    iv = parse_cuda(image_cuda)
    mv = parse_cuda(min_cuda)
    if iv is None or mv is None:
        return True
    return iv >= mv
